spanner/benchmark/ycsb.py

```diff
@@ -0,0 +1,257 @@
+    latency_array = numpy.array(latencies_ms[operation])
+    for j in range(num_bucket):
+      print '[%s], %d, %d' % (
+          operation_upper, j,
+          ((j <= latency_array) & (latency_array < (j + 1))).sum())
+    print '[%s], >%d, %d' % (
+          operation_upper, num_bucket, (num_bucket <= latency_array).sum())
+
+
+class WorkloadThread(threading.Thread):
+  """A single thread running workload."""
+
+  def __init__(self, database, parameters, total_weight, weights, operations):
+    threading.Thread.__init__(self)
+    self._database = database
+    self._parameters = parameters
+    self._total_weight = total_weight
+    self._weights = weights
+    self._operations = operations
+    self._latencies_ms = {}
+    for operation in self._operations:
+      self._latencies_ms[operation] = []
+
+  def run(self):
+    """Run a single thread of the workload."""
+    i = 0
+    operation_count = int(self._parameters['operationcount'])
+    while i < operation_count:
+      i += 1
+      weight = random.uniform(0, self._total_weight)
+      for j in range(len(self._weights)):
+        if weight <= self._weights[j]:
+          DoOperation(self._database, self._parameters['table'],
+                      self._operations[j], self._latencies_ms)
+          break
+
+  def latencies_ms(self):
+    """Returns the latencies."""
+    return self._latencies_ms
+
+
+def run_workload(database, parameters):
+  """Runs workload against the database."""
+  total_weight = 0.0
+  weights = []
+  operations = []
+  latencies_ms = {}
+  for operation in OPERATIONS:
+    weight = float(parameters[operation])
+    if weight <= 0.0:
+      continue
+    total_weight += weight
+    op_code = operation.split('proportion')[0]
+    operations.append(op_code)
+    weights.append(total_weight)
+    latencies_ms[op_code] = []
+
+  threads = []
+  start = timeit.default_timer()
+  for i in range(int(parameters['num_worker'])):
+    thread = WorkloadThread(database, parameters, total_weight, weights,
+                            operations)
+    thread.start()
+    threads.append(thread)
+
+  for thread in threads:
+    thread.join()
+  end = timeit.default_timer()
+
+  for thread in threads:
+    thread_latencies_ms = thread.latencies_ms()
+    for key in latencies_ms.keys():
+      latencies_ms[key].extend(thread_latencies_ms[key])
+
+  aggregate_metrics(latencies_ms, (end - start) * 1000.0,
+                    parameters['num_bucket'])
+
+
+if __name__ == '__main__':
+  parameters = parse_options()
+  if parameters['command'] == 'run':
+    if 'cloudspanner.channels' in parameters:
+      assert parameters['cloudspanner.channels'] == 1, ('Python doesn\'t '
+                                                        'support channels > 1.')
+    database = open_database(parameters)
+    load_keys(database, parameters)
+    run_workload(database, parameters)
+  else:
+    raise Exception('Unknown command %s.' % parameters['command'])
```

Comment:
Same here, more specific exception?

I think that's all of the comments I have on the things I know.  I'll pass this on to @tseaver if he has other comments and @vkedia and @tseaver on the GIL part.

Reply:
Thx.